diff --git a/minuitwrp/graphics.cpp b/minuitwrp/graphics.cpp
index 47325d0..cd9e9f6 100644
--- a/minuitwrp/graphics.cpp
+++ b/minuitwrp/graphics.cpp
@@ -292,7 +292,11 @@ static void get_memory_surface(GGLSurface* ms) {
     ms->format = gr_draw->format;
 }
 
+#ifdef STE_HARDWARE
+int gr_init_real(void)
+#else
 int gr_init(void)
+#endif
 {
     gr_draw = NULL;
 
@@ -363,6 +367,27 @@ int gr_init(void)
     return 0;
 }
 
+#ifdef STE_HARDWARE
+/*
+ * FIXME: This is a total hack.
+ *
+ * Round 1
+ * framebuffer: fd 4 (480 x 800)
+ *
+ * Round 2
+ * framebuffer: fd 6 (480 x 800)
+ *
+*/
+int gr_init(void) {
+    int ret;
+    // Round 1
+    gr_init_real();
+    // Return the result of round 2
+    ret = gr_init_real();
+    return ret;
+}
+#endif
+
 void gr_exit(void)
 {
     gr_backend->exit(gr_backend);
-- 
2.5.0

