diff --git a/linker/linker.cpp b/linker/linker.cpp
index 58fa79e..a86e081 100644
--- a/linker/linker.cpp
+++ b/linker/linker.cpp
@@ -2223,8 +2223,6 @@ bool soinfo::LinkImage(const android_dlextinfo* extinfo) {
   if (has_text_relocations) {
     // Make segments writable to allow text relocations to work properly. We will later call
     // phdr_table_protect_segments() after all of them are applied and all constructors are run.
-    DL_WARN("%s has text relocations. This is wasting memory and prevents "
-            "security hardening. Please fix.", name);
     if (phdr_table_unprotect_segments(phdr, phnum, load_bias) < 0) {
       DL_ERR("can't unprotect loadable segments for \"%s\": %s",
              name, strerror(errno));
-- 
2.1.0

