diff --git a/vibrator/Android.mk b/vibrator/Android.mk
index 6681f84..84d73c8 100644
--- a/vibrator/Android.mk
+++ b/vibrator/Android.mk
@@ -1,4 +1,9 @@
 # Copyright 2006 The Android Open Source Project
 
-LOCAL_SRC_FILES += vibrator/vibrator.c
-
+## Must point to a source file that implements the sendit() function
+ ifneq ($(BOARD_HAS_VIBRATOR_IMPLEMENTATION),)
+     LOCAL_SRC_FILES += $(BOARD_HAS_VIBRATOR_IMPLEMENTATION)
+     LOCAL_CFLAGS += -DUSE_ALTERNATIVE_VIBRATOR
+else
+	LOCAL_SRC_FILES += vibrator/vibrator.c
+endif
-- 
2.1.0

