diff --git a/tools/releasetools/common.py b/tools/releasetools/common.py
index 499e8c2..ca054dd 100755
--- a/tools/releasetools/common.py
+++ b/tools/releasetools/common.py
@@ -1258,6 +1258,10 @@ def ParseCertificate(data):
 
 def MakeRecoveryPatch(input_dir, output_sink, recovery_img, boot_img,
                       info_dict=None):
+  pass
+
+def MakeRecoveryPatch2(input_dir, output_sink, recovery_img, boot_img,
+                      info_dict=None):
   """Generate a binary patch that creates the recovery image starting
   with the boot image.  (Most of the space in these images is just the
   kernel, which is identical for the two, so the resulting patch
-- 
2.1.0

